--- build/env.cmake
+++ build/env.cmake
@@ -160,7 +160,9 @@ endif()
 add_compile_definitions( _ARCH_BITS=${BITS} ${ARCH} )
 
 # global compiler warnings settings
+if(NOT _NCBIVDB_CFG_PACKAGING)
 add_definitions( -Wall )
+endif()
 if ( "GNU" STREQUAL "${CMAKE_C_COMPILER_ID}")
 elseif ( CMAKE_CXX_COMPILER_ID MATCHES "^(Apple)?Clang$" )
 elseif ( "MSVC" STREQUAL "${CMAKE_C_COMPILER_ID}")
@@ -386,8 +388,8 @@ if( RUN_SANITIZER_TESTS_OVERRIDE )
 endif()
 message( "RUN_SANITIZER_TESTS: ${RUN_SANITIZER_TESTS}" )
 
+endif(NOT _NCBIVDB_CFG_PACKAGING)
 if( NOT HAVE_MBEDTLS_F )
 	message( "Using local mbedtls headers from interfaces/ext/mbedtls" )
 	include_directories( interfaces/ext/mbedtls )
 endif()
-endif(NOT _NCBIVDB_CFG_PACKAGING)
